src/org/javarosa/core/model/ItemsetBinding.java

```diff
@@ -32,31 +42,37 @@
      * xform/xpath classes, which we don't from the core model project
      */
 
-    public TreeReference nodesetRef;   //absolute ref of itemset source nodes
     public IConditionExpr nodesetExpr; //path expression for source nodes; may be relative, may contain predicates
     public TreeReference contextRef;   //context ref for nodesetExpr; ref of the control parent (group/formdef) of itemset question
-       //note: this is only here because its currently impossible to both (a) get a form control's parent, and (b)
-       //convert expressions into refs while preserving predicates. once these are fixed, this field can go away
+    //note: this is only here because its currently impossible to both (a) get a form control's parent, and (b)
+    //convert expressions into refs while preserving predicates. once these are fixed, this field can go away
 
-    public TreeReference labelRef;     //absolute ref of label
     public IConditionExpr labelExpr;   //path expression for label; may be relative, no predicates
     public boolean labelIsItext;       //if true, content of 'label' is an itext id
 
     public boolean copyMode;           //true = copy subtree; false = copy string value
     public IConditionExpr copyExpr;    // path expression for copy; may be relative, no predicates
-    public TreeReference copyRef;      //absolute ref to copy
-    public TreeReference valueRef;     //absolute ref to value
     public IConditionExpr valueExpr;   //path expression for value; may be relative, no predicates (must be relative if copy mode)
 
     private TreeReference destRef; //ref that identifies the repeated nodes resulting from this itemset
-                                   //not serialized -- set by QuestionDef.setDynamicChoices()
+    //not serialized -- set by QuestionDef.setDynamicChoices()
     private List<SelectChoice> choices; //dynamic choices -- not serialized, obviously
 
-    public List<SelectChoice> getChoices () {
+    private DataInstance instanceModel;
+    private Optional<XPathNodeset> nodeset;
```

Comment:
Will [Optional](https://developer.android.com/reference/java/util/Optional)—added in API level 24—work in Collect?
```
    defaultConfig {
        applicationId('org.odk.collect.android')
        minSdkVersion(16)
        targetSdkVersion(22)
```

Reply:
Yes, `targetSdk` just determines forwards-compatibility UI features. We're still compiling with API 26.

Reply:
I said that with high confidence and then I realized that's not true for Android features. Like I said in another comment, I'm not really sure how all that interacts with pure Java dependencies. I tried to add a jar with this code to Collect and it does compile.

Reply:
Hmm, seems possible.

Reply:
Did you get an answer to Why Guava Optional over Java 8?
https://github.com/opendatakit/collect/pull/1785#pullrequestreview-89992396

Reply:
Well, here’s what it looks like:

```
05-05 23:52:44.902 21563-21696/org.odk.collect.android E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #2
                                                                         Process: org.odk.collect.android, PID: 21563
                                                                         java.lang.RuntimeException: An error occurred while executing doInBackground()
                                                                             at android.os.AsyncTask$3.done(AsyncTask.java:309)
                                                                             at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
                                                                             at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
                                                                             at java.util.concurrent.FutureTask.run(FutureTask.java:242)
                                                                             at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
                                                                             at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                             at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                             at java.lang.Thread.run(Thread.java:818)
                                                                          Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/Optional;
```

Reply:
Maybe look at Guava’s Optional?

Reply:
I don't know... We would have to add the dependency and, eventually we would have to remove it to migrate to native Optionals... it feels like too much just to use Optional in two classes. I think we can wait.

Reply:
Makes sense.

Reply:
Alright, just pushed changes from your branch and removed Optional too

Reply:
In that case I’ll have another go with Collect (unless you are already ahead of me with the problem solving).

Reply:
Please, go ahead :)

Reply:
Darn. ;-)

Reply:
You’re still using Stream. :-(

Reply:
None of your modern sorcery in Collect! ;-)
```
    public void initReferences(DataInstance instanceModel, List<DataInstance> nonMainInstances, QuestionDef q) {

        Stream<DataInstance> stream = nonMainInstances.stream();
```

Reply:
Do we have to remove that too? :D

Reply:
I’ll give you the answer in the form of a stack trace.
```
Caused by: java.lang.NoSuchMethodError: No interface method stream()Ljava/util/stream/Stream; in class Ljava/util/List; or its super classes (declaration of 'java.util.List' appears in /system/framework/core-libart.jar)
                                                                             at org.javarosa.core.model.ItemsetBinding.initReferences(ItemsetBinding.java:120)
                                                                         
```

Reply:
Is it OK to use Stream in tests?

Reply:
Nevermind, I'll get rid of those too

Reply:
I assume so, since this is a runtime error.